File: lib/scripts/select-repos.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const tty = require("tty");
const { discover } = require("./discover-repos");

const CONTINUE_ROW = 0;
const MANUAL_ROW = 1;

function parseArgs(argv) {
  let workspaceRoot = "";
  let outputPath = "";
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--output") {
      outputPath = argv[index + 1] || "";
      index += 1;
      continue;
    }
    if (!workspaceRoot) {
      workspaceRoot = arg;
    }
  }
  return { workspaceRoot, outputPath };
}

function openInputStream() {
  if (process.stdin.isTTY) {
    return process.stdin;
  }
  try {
    const fd = fs.openSync("/dev/tty", "r");
    return new tty.ReadStream(fd);
  } catch {
    return null;
  }
}

function canRunInteractive() {
  if (process.stdin.isTTY || process.stderr.isTTY) {
    return true;
  }
  try {
    const fd = fs.openSync("/dev/tty", "r");
    fs.closeSync(fd);
    return true;
  } catch {
    return false;
  }
}

function formatRepoLine(repo, selected, active) {
  const marker = active ? "›" : " ";
  const circle = selected ? "●" : "○";
  return `${marker} ${circle} ${repo.name}  (${repo.default_branch}, ${repo.runtime_profile})`;
}

function formatFooterLine(label, active) {
  const marker = active ? "›" : " ";
  return `${marker} ${label}`;
}

function render(workspaceRoot, repos, selected, cursor, footerCursor, statusMessage) {
  const lines = [];
  lines.push(`Repositories under ${workspaceRoot}`);
  lines.push("");
  repos.forEach((repo, index) => {
    const active = footerCursor === null && cursor === index;
    lines.push(formatRepoLine(repo, selected.has(index), active));
  });
  lines.push("");
  const count = selected.size;
  const noun = count === 1 ? "repository" : "repositories";
  const continueLabel = `Continue with ${count} selected ${noun}`;
  lines.push(formatFooterLine(continueLabel, footerCursor === CONTINUE_ROW));
  lines.push(formatFooterLine("Add repository manually", footerCursor === MANUAL_ROW));
  lines.push("");
  lines.push("↑↓ move   Space toggle   Enter confirm   a select all   q cancel");
  if (statusMessage) {
    lines.push(statusMessage);
  }
  process.stderr.write(`\x1b[?25l\x1b[2J\x1b[H${lines.join("\n")}`);
}

function restoreTerminal(input) {
  process.stderr.write("\x1b[?25h\x1b[0J\n");
  if (input.isTTY) {
    input.setRawMode(false);
  }
}

function writeResult(result, outputPath) {
  const json = `${JSON.stringify(result, null, 2)}\n`;
  if (outputPath) {
    fs.writeFileSync(outputPath, json, "utf8");
    return;
  }
  process.stdout.write(json);
}

function runInteractive(workspaceRoot, repos, outputPath) {
  const input = openInputStream();
  if (!input) {
    process.stderr.write("Interactive repository selection requires a TTY.\n");
    process.exit(1);
  }

  const selected = new Set();
  let cursor = 0;
  let footerCursor = null;
  let statusMessage = "";

  const draw = () => render(workspaceRoot, repos, selected, cursor, footerCursor, statusMessage);

  const finish = (result, code) => {
    restoreTerminal(input);
    input.pause();
    input.removeAllListeners("data");
    if (code === 0 && result) {
      writeResult(result, outputPath);
    }
    process.exit(code);
  };

  if (input.isTTY) {
    input.setRawMode(true);
  }
  input.resume();
  input.setEncoding("utf8");

  draw();

  input.on("data", (chunk) => {
    statusMessage = "";

    if (chunk === "\u0003" || chunk === "q") {
      finish(null, 130);
    }

    if (chunk === "a") {
      if (selected.size === repos.length) {
        selected.clear();
      } else {
        for (let index = 0; index < repos.length; index += 1) {
          selected.add(index);
        }
      }
      draw();
      return;
    }

    if (chunk === "\u001b[A") {
      if (footerCursor !== null) {
        if (footerCursor === MANUAL_ROW) {
          footerCursor = CONTINUE_ROW;
        } else {
          footerCursor = null;
          cursor = repos.length - 1;
        }
      } else if (cursor > 0) {
        cursor -= 1;
      }
      draw();
      return;
    }

    if (chunk === "\u001b[B") {
      if (footerCursor !== null) {
        if (footerCursor === CONTINUE_ROW) {
          footerCursor = MANUAL_ROW;
        }
      } else if (cursor < repos.length - 1) {
        cursor += 1;
      } else {
        footerCursor = CONTINUE_ROW;
      }
      draw();
      return;
    }

    if (chunk === " ") {
      if (footerCursor === null) {
        if (selected.has(cursor)) {
          selected.delete(cursor);
        } else {
          selected.add(cursor);
        }
        draw();
      }
      return;
    }

    if (chunk === "\r" || chunk === "\n") {
      if (footerCursor === MANUAL_ROW) {
        finish(null, 2);
        return;
      }
      if (selected.size === 0) {
        statusMessage = "Select at least one repository (Space to toggle), or choose manual entry.";
        draw();
        return;
      }
      const result = [...selected].sort((a, b) => a - b).map((index) => repos[index]);
      finish(result, 0);
      return;
    }

    if (chunk.startsWith("\u001b")) {
      return;
    }
  });
}

function usage() {
  process.stderr.write("Usage: select-repos.js <workspace-root> [--output <file>]\n");
  process.exit(1);
}

function main() {
  const { workspaceRoot, outputPath } = parseArgs(process.argv.slice(2));
  if (!workspaceRoot) {
    usage();
  }

  if (!canRunInteractive()) {
    process.stderr.write("Interactive repository selection requires a TTY.\n");
    process.exit(1);
  }

  const repos = discover(workspaceRoot);
  if (repos.length === 0) {
    process.stderr.write("No repositories found.\n");
    process.exit(1);
  }

  runInteractive(workspaceRoot, repos, outputPath);
}

main();

File: lib/scripts/select_repos.py
#!/usr/bin/env python3
"""Interactive repository picker for Lumen init."""

from __future__ import annotations

import json
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
if str(SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPT_DIR))

from discover_repos import discover, parse_selection


def parse_args(argv: list) -> tuple:
    workspace_root = ""
    output_path = ""
    index = 0
    while index < len(argv):
        arg = argv[index]
        if arg == "--output":
            output_path = argv[index + 1] if index + 1 < len(argv) else ""
            index += 2
            continue
        if not workspace_root:
            workspace_root = arg
        index += 1
    return workspace_root, output_path


def can_run_interactive() -> bool:
    if sys.stdin.isatty() or sys.stderr.isatty():
        return True
    try:
        handle = open("/dev/tty", "rb")
        handle.close()
        return True
    except OSError:
        return False


def prompt_line(message: str) -> str:
    sys.stderr.write(message)
    sys.stderr.flush()
    answer = sys.stdin.readline()
    if answer is None:
        return ""
    return answer.strip()


def write_result(result: list, output_path: str) -> None:
    payload = json.dumps(result, indent=2) + "\n"
    if output_path:
        Path(output_path).write_text(payload, encoding="utf-8")
        return
    sys.stdout.write(payload)


def print_repo_menu(workspace_root: str, repos: list, selected: set) -> None:
    sys.stderr.write(f"\nRepositories under {workspace_root}\n\n")
    for index, repo in enumerate(repos, start=1):
        mark = "x" if index - 1 in selected else " "
        sys.stderr.write(
            f"  [{mark}] {index}) {repo['name']} "
            f"({repo['default_branch']}, {repo['runtime_profile']})\n"
        )
    count = len(selected)
    noun = "repository" if count == 1 else "repositories"
    sys.stderr.write(f"\nSelected: {count} {noun}\n\n")
    sys.stderr.write("  1,3,5   toggle repositories by number\n")
    sys.stderr.write("  all     select all\n")
    sys.stderr.write("  clear   clear selection\n")
    sys.stderr.write("  manual  add repository manually\n")
    sys.stderr.write("  Enter   continue with selected repositories\n")
    sys.stderr.write("  q       cancel\n\n")
    sys.stderr.flush()


def apply_toggle(selected: set, repos: list, answer: str) -> bool:
    lowered = answer.strip().lower()
    if lowered in {"all", "a"}:
        selected.clear()
        selected.update(range(len(repos)))
        return True
    if lowered in {"clear", "c", "none"}:
        selected.clear()
        return True

    try:
        indices = parse_selection(answer, len(repos))
    except RuntimeError as exc:
        sys.stderr.write(f"{exc}\n")
        return False

    for index in indices:
        zero_based = index - 1
        if zero_based in selected:
            selected.remove(zero_based)
        else:
            selected.add(zero_based)
    return True


def run_interactive(workspace_root: str, repos: list, output_path: str) -> int:
    selected: set = set()

    while True:
        print_repo_menu(workspace_root, repos, selected)
        answer = prompt_line("Selection: ")

        if answer.lower() in {"q", "quit", "cancel"}:
            return 130
        if answer.lower() in {"manual", "m", "+"}:
            return 2
        if answer == "":
            if not selected:
                sys.stderr.write("Select at least one repository, or type manual to add one by URL.\n")
                continue
            result = [repos[index] for index in sorted(selected)]
            write_result(result, output_path)
            return 0

        apply_toggle(selected, repos, answer)


def usage() -> None:
    sys.stderr.write("Usage: select_repos.py <workspace-root> [--output <file>]\n")
    raise SystemExit(1)


def main() -> int:
    workspace_root, output_path = parse_args(sys.argv[1:])
    if not workspace_root:
        usage()

    if not can_run_interactive():
        sys.stderr.write("Interactive repository selection requires a TTY.\n")
        return 1

    repos = discover(workspace_root)
    if not repos:
        sys.stderr.write("No repositories found.\n")
        return 1

    return run_interactive(workspace_root, repos, output_path)


if __name__ == "__main__":
    raise SystemExit(main())
